src/async/GenericMuxedInterruptDistributor.cc

```diff
@@ -3,27 +3,718 @@
+  inline void GenericMuxedInterruptDistributor::enableOneInterrupt(uint32_t ithInterrupt) {
+    enableInterruptsFromMask(iToMask(ithInterrupt));
+  }
+
+  /********************************************************************************************************************/
   void GenericMuxedInterruptDistributor::handle(VersionNumber version) {
-    // Stupid testing implementation that always triggers all children
+    try {
+      _isr->read();
+      uint32_t ipr = _activeInterrupts & _isr->accessData(0);
+
+      for(auto const& [i, subDomainWeakPtr] : _subDomains) {
+        if(ipr & iToMask(i)) {
+          if(auto subDomain = subDomainWeakPtr.lock(); subDomain) {
+            //  The weak pointer might have gone.
+            //  TODO FIXME: We need a cleanup function which removes the map entry.
+            //  Otherwise we might be stuck with a bad weak pointer which is tried in each handle() call.
+
+            subDomain->distribute(nullptr, version);
+
+            // Requirement: nested interrupt handlers must clear their active interrupt flag first,
+            // then the parent interrupt flags are cleared.
+            clearOneInterrupt(i);
+          }
+        }
+      } // for
+    }
+    catch(ChimeraTK::runtime_error&) {
+      // There's nothing to do. The transferElement part of _activeInterrupts has already called the backend's setException
+    }
+  } // handle
+
+  /********************************************************************************************************************/
+  std::unique_ptr<GenericMuxedInterruptDistributor> GenericMuxedInterruptDistributor::create(
+      [[maybe_unused]] std::string const& description, boost::shared_ptr<SubDomain<std::nullptr_t>> parent) {
+    /*
+     *  This is a factory function. It parses the json, and calls the constructor.
+     *  It returns an initalized GenericMuxedInterruptDistributor.
+     *  'descriptor' is a JSON snippet containing configuration data
+     */
+
+    auto parseResult = parseAndValidateJsonDescriptionStrV0(parent->getId(), description);
+    std::bitset<OPTION_CODE_COUNT> optionRegisterSettings = parseResult.first;
+    std::string registerPath = parseResult.second;
+
+    return std::make_unique<GenericMuxedInterruptDistributor>(std::move(parent), registerPath, optionRegisterSettings);
+  } // create
+
+  /********************************************************************************************************************/
+  void GenericMuxedInterruptDistributor::activate(VersionNumber version) {
+    if(_hasMer) { // Set MER: turn on the Master Enable and HIE (hardware interrupt enable) bits
+      try {
+        _mer->accessData(0) = 0x00000003;
+        _mer->write();
+      }
+      catch(ChimeraTK::runtime_error&) {
+      }
+    }
+
+    // Disable any interrupts for which there is no valid subDomain.
+    uint32_t activeInterrupts = 0;
+    for(auto const& [i, subDomainWeakPtr] : _subDomains) {
+      try {
```

Comment:
Is i the index of the subdomain in the bitset?

Reply:
Yes, I made a code comment.